### ISSUE TITLE
Give more helpful error when shadowing `const`

### ISSUE THREAD

**Issue by IGI-111** [FuelLabs/sway#4587]
Shadowing constants is forbidden, but the compiler will currently give an unhelpful error message like so:
 ```
 The name `VAL` is defined multiple times
 ```
We should give a clearer error message in that particular case.

**Comment by deepak-likes-code** [comment]
I can pick this up

**Comment by ironcev** [comment]
Hi, this looks like a _great_ first issue, not only a good one :-) I would love to contribute to it. I have two questions, before starting with the implementation. This is the TL;DR version of the questions. I will also provide a longer one in the comment below. The longer one is more for me, and as a checklist once the feature is implemented. I've started diving deeper into Sway to get a better feeling for the language and the design decisions already done. So I systematically piled up all the cases realted to the shadowing `const`s.
 
## TL;DR

(The terms _item-style shadowing_ and _sequential shadowing_ have the meaning defined in the [Rust RFC 116 (No module shadowing)](https://rust-lang.github.io/rfcs/0116-no-module-shadowing.html).)

:question: Should the new message apply only to _sequential shadowing_, and for the _item-style shadowing_ we keep the existing one?

This is how I understand the motivation behind this issue - to give hint to the programmer that unlike variables `const`s cannot be shadowed on the same or nested scope level.

Having a separate error message for `const` in case of item-style shadowing makes no sense for me, because I don't see why would `const` be an exceptional item and deserve an error message different then other items.

Applied to the [shadowed consts and vars tests](https://github.com/FuelLabs/sway/blob/ac4ff14482471804d80faffe659333f10fa421e5/test/src/e2e_vm_tests/test_programs/should_fail/shadowing/shadowed_consts_and_vars/src/main.sw) this would mean the following test output:

    // const shadowing an imported const
    use lib::X;
    const X = 6; // <existing error> The name `X` is defined multiple times
 
    // const shadowing a local const
    const Y = 7;
    const Y = 8; // <existing error> The name `X` is defined multiple times
    fn main() {
        // variable shadowing an imported const
        let X = 9; // <new error>
 
        // variable shadowing a const
        const Z = 3;
        let Z = 4; // <new error>
 
        // const shadowing a variable
        let W = 2;
        const W = 1; // <new error>
 
        // Variable shadowing a variable - this is okay!
        let P = 7;
        let P = 8;
    }

The detailed reasoning is in the comment below.

:question: How should the new error message exactly look like?                                                                                               

I have difficulties formulating a good single message that would cover all three cases. E.g.:
    
    The name `VAL` shadows a constant or a variable with the same name.

Or should we go for three different error messages? E.g.:

    The constant `VAL` shadows the constant of the same name.
    The constant `VAL` shadows the variable of the same name.
    The variable `VAL` shadows the constant of the same name.

**Comment by IGI-111** [comment]
I think the reasoning here is sound, the new error seems indeed most appropriate for sequential shadowing.

I think it's probably clearer to go with distinct error messages for each case, but It might also be worth it restating the rule that constants cannot be shadowed as part of each message so as to make it clear why this is erroneous.

**Comment by ironcev** [comment]
Another one edge-case to consider. `use` importing has completely different error message at the moment:

    The name "LIB_X" shadows another symbol with the same name.

In the case of _item-style shadowing_ this message should stay the same. But when we `use` identifiers within functions, importing acts sequentially. Applying the above logic we would have the following:

    use libA::X;
    use libB::X; // <existing error> The name "X" shadows another symbol with the same name.

    fn f() -> () {
        use libC::X;
        use libD::X; // <new error>
    }

Does this sound good or I missed some detail? 

Side note: we already have a few similar error messages for multiple definitions on names, specific messages for multiple imported definitions, an even a specific error message for multiple definitions of constants ;-) I see a potential for a good clean up here. I'll open a separate issue for that.

**Comment by ironcev** [comment]
It turned out that in also in this case, there is no error message at the moment for this case, same like in the comment above with `let`s:

```
const X: u32 = 1;
fn f() -> () {
    const X: u32 = 2; // <new error>
    {
        const X: u32 = 3; // !! Currently no error message here.
     }
}
```
Since this issue deals only with turning the existing error messages into more helpful once, we will keep not having the error in this case.
